### PR TITLE
ARROW-17786: [Java] Read CSV files using org.apache.arrow.dataset.jni.NativeDatasetFactory

### PR DESCRIPTION
**Pull request by davisusanibar**
Support CSV file format in java Dataset API

### CODE REVIEW

File: java/dataset/src/test/java/org/apache/arrow/dataset/CsvWriteSupport.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import java.util.Random;
+
+public class CsvWriteSupport {
+  private final String path;
+  private final String uri;
+  private final Random random = new Random();
+
+  public CsvWriteSupport(File outputFolder) {
+    path = outputFolder.getPath() + File.separator + "generated-" + random.nextLong() + ".csv";
+    uri = "file://" + path;
```
Comment (pitrou) [inline java/dataset/src/test/java/org/apache/arrow/dataset/CsvWriteSupport.java:34]: Shouldn't you use a URI constructor instead?
https://docs.oracle.com/javase/7/docs/api/java/net/URI.html#URI(java.lang.String,%20java.lang.String,%20java.lang.String)

**Comment by pitrou** [comment]
@lwhite1 Would you like to review this?

**Comment by pitrou** [comment]
@davisusanibar Why is only write support tested? Shouldn't you test support for dataset reads as well?

**Comment by davisusanibar** [comment]
> @davisusanibar Why is only write support tested? Shouldn't you test support for dataset reads as well?

Hi, write support is doing by Java native libraries, then Dataset module is reading that CSV file with (FileFormat.CSV): `new FileSystemDatasetFactory(rootAllocator(), NativeMemoryPool.getDefault(), FileFormat.CSV, writeSupport.getOutputURI());`

Currently we are only offering Read support for Parquet. ORC and now CSV, ... write support it not implemented

**Comment by pitrou** [comment]
> > @davisusanibar Why is only write support tested? Shouldn't you test support for dataset reads as well?
> 
> Hi, write support is doing by Java native libraries, then Dataset module is reading that CSV file with (FileFormat.CSV): `new FileSystemDatasetFactory(rootAllocator(), NativeMemoryPool.getDefault(), FileFormat.CSV, writeSupport.getOutputURI());`
> 
> Currently we are only offering Read support for Parquet. ORC and now CSV, ... write support it not implemented

Oops, sorry, I had misread the tests.

File: java/dataset/src/test/java/org/apache/arrow/dataset/CsvWriteSupport.java (hunk 1 above)
Comment (davisusanibar) [reply java/dataset/src/test/java/org/apache/arrow/dataset/CsvWriteSupport.java]: Thanks, changed

**Comment by davisusanibar** [comment]
This line https://github.com/apache/arrow/blob/91e3ac53e2e21736ce6291d73fc37da6fa21259d/cpp/src/arrow/json/parser.cc#L1030 was not changed long time ago, I don't know what is the error: 
````
[1/1] cd /tmp/arrow-lint-5g90mrbe/cpp-build && /usr/local/bin/python /arrow/cpp/build-support/run_cpplint.py --cpplint_binary /arrow/cpp/build-support/cpplint.py --exclude_globs /arrow/cpp/build-support/lint_exclusions.txt --source_dir /arrow/cpp/src --source_dir /arrow/cpp/examples --source_dir /arrow/cpp/tools --quiet
FAILED: CMakeFiles/lint 
cd /tmp/arrow-lint-5g90mrbe/cpp-build && /usr/local/bin/python /arrow/cpp/build-support/run_cpplint.py --cpplint_binary /arrow/cpp/build-support/cpplint.py --exclude_globs /arrow/cpp/build-support/lint_exclusions.txt --source_dir /arrow/cpp/src --source_dir /arrow/cpp/examples --source_dir /arrow/cpp/tools --quiet
/arrow/cpp/src/arrow/json/parser.cc:1030:  Lines should be <= 90 characters long  [whitespace/line_length] [2]
Done processing /arrow/cpp/src/arrow/json/parser.cc
Total errors found: 1
````

**Comment by davisusanibar** [comment]
> This line
> 
> https://github.com/apache/arrow/blob/91e3ac53e2e21736ce6291d73fc37da6fa21259d/cpp/src/arrow/json/parser.cc#L1030
> 
> was not changed long time ago, I don't know what is the error:
> ```
> [1/1] cd /tmp/arrow-lint-5g90mrbe/cpp-build && /usr/local/bin/python /arrow/cpp/build-support/run_cpplint.py --cpplint_binary /arrow/cpp/build-support/cpplint.py --exclude_globs /arrow/cpp/build-support/lint_exclusions.txt --source_dir /arrow/cpp/src --source_dir /arrow/cpp/examples --source_dir /arrow/cpp/tools --quiet
> FAILED: CMakeFiles/lint 
> cd /tmp/arrow-lint-5g90mrbe/cpp-build && /usr/local/bin/python /arrow/cpp/build-support/run_cpplint.py --cpplint_binary /arrow/cpp/build-support/cpplint.py --exclude_globs /arrow/cpp/build-support/lint_exclusions.txt --source_dir /arrow/cpp/src --source_dir /arrow/cpp/examples --source_dir /arrow/cpp/tools --quiet
> /arrow/cpp/src/arrow/json/parser.cc:1030:  Lines should be <= 90 characters long  [whitespace/line_length] [2]
> Done processing /arrow/cpp/src/arrow/json/parser.cc
> Total errors found: 1
> ```

Just rebase and working fine now

**Comment by davisusanibar** [comment]
Please, when you have some time if you could review this, thank you

File: java/dataset/src/test/java/org/apache/arrow/dataset/file/TestFileSystemDataset.java [hunk 2]
```diff
@@ -361,6 +362,30 @@ public void testBaseOrcRead() throws Exception {
...
+
+    System.out.println(schema);
+    assertScanBatchesProduced(factory, options);
+    assertEquals(1, datum.size());
+    assertEquals(2, schema.getFields().size());
+    assertEquals("Name", schema.getFields().get(0).getName());
+
+    checkParquetReadResult(schema, expectedJsonUnordered, datum);
+
+    AutoCloseables.close(datum);
```
Comment (lidavidm) [inline java/dataset/src/test/java/org/apache/arrow/dataset/file/TestFileSystemDataset.java:385]: Use try-with-resources?

**Comment by lwhite1** [comment]
Is this for release 10?

File: java/dataset/src/main/cpp/jni_wrapper.cc [hunk 3]
```diff
@@ -91,10 +91,10 @@ arrow::Result<std::shared_ptr<arrow::dataset::FileFormat>> GetFileFormat(
       return std::make_shared<arrow::dataset::ParquetFileFormat>();
     case 1:
       return std::make_shared<arrow::dataset::IpcFileFormat>();
-#ifdef ARROW_ORC
```
[the remark this answers is not in the document]
Comment (davisusanibar) [reply java/dataset/src/main/cpp/jni_wrapper.cc:94]: Added

File: java/dataset/src/test/java/org/apache/arrow/dataset/file/TestFileSystemDataset.java (hunk 2 above)
Comment (davisusanibar) [reply java/dataset/src/test/java/org/apache/arrow/dataset/file/TestFileSystemDataset.java]: Added try-with for `FileSystemDatasetFactory` but don't know how to add that for`List<ArrowRecordBatch>` that is an input for another function.

**Comment by davisusanibar** [comment]
> Sorry for not getting to this sooner. Is this for release 10?

Not problem for that. Yes, trying to merge for 10 to then could create cookbooks for this CSV Reader case

**Comment by lidavidm** [comment]
@github-actions crossbow submit *java*

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
CI shows a flake in testTable:

```
Error:  org.apache.arrow.c.RoundtripTest.testTable  Time elapsed: 0.031 s  <<< ERROR!
java.lang.IllegalStateException: Cannot import released ArrowSchema
	at org.apache.arrow.util.Preconditions.checkState(Preconditions.java:458)
	at org.apache.arrow.c.SchemaImporter.importField(SchemaImporter.java:63)
	at org.apache.arrow.c.SchemaImporter.importField(SchemaImporter.java:56)
	at org.apache.arrow.c.Data.importField(Data.java:246)
	at org.apache.arrow.c.Data.importSchema(Data.java:266)
	at org.apache.arrow.c.Data.importVectorSchemaRoot(Data.java:377)
	at org.apache.arrow.c.RoundtripTest.testTable(RoundtripTest.java:683)
```

**Comment by lidavidm** [comment]
> CI shows a flake in testTable:
> 
> ```
> Error:  org.apache.arrow.c.RoundtripTest.testTable  Time elapsed: 0.031 s  <<< ERROR!
> java.lang.IllegalStateException: Cannot import released ArrowSchema
> 	at org.apache.arrow.util.Preconditions.checkState(Preconditions.java:458)
> 	at org.apache.arrow.c.SchemaImporter.importField(SchemaImporter.java:63)
> 	at org.apache.arrow.c.SchemaImporter.importField(SchemaImporter.java:56)
> 	at org.apache.arrow.c.Data.importField(Data.java:246)
> 	at org.apache.arrow.c.Data.importSchema(Data.java:266)
> 	at org.apache.arrow.c.Data.importVectorSchemaRoot(Data.java:377)
> 	at org.apache.arrow.c.RoundtripTest.testTable(RoundtripTest.java:683)
> ```

@lwhite1 have you seen this? Might be worth some investigation

**Comment by lidavidm** [comment]
@davisusanibar it may be worth investigating if we can not build/run tests in java-jars? It seems rather redundant (or maybe that's on purpose)

**Comment by davisusanibar** [comment]
> @davisusanibar it may be worth investigating if we can not build/run tests in java-jars? It seems rather redundant (or maybe that's on purpose)

Yes, sound redundant at first review, but, let me check what part could be improved

**Comment by lwhite1** [comment]
> > CI shows a flake in testTable:
> > ```
> > Error:  org.apache.arrow.c.RoundtripTest.testTable  Time elapsed: 0.031 s  <<< ERROR!
> > java.lang.IllegalStateException: Cannot import released ArrowSchema
> > 	at org.apache.arrow.util.Preconditions.checkState(Preconditions.java:458)
> > 	at org.apache.arrow.c.SchemaImporter.importField(SchemaImporter.java:63)
> > 	at org.apache.arrow.c.SchemaImporter.importField(SchemaImporter.java:56)
> > 	at org.apache.arrow.c.Data.importField(Data.java:246)
> > 	at org.apache.arrow.c.Data.importSchema(Data.java:266)
> > 	at org.apache.arrow.c.Data.importVectorSchemaRoot(Data.java:377)
> > 	at org.apache.arrow.c.RoundtripTest.testTable(RoundtripTest.java:683)
> > ```
> 
> @lwhite1 have you seen this? Might be worth some investigation
> 
I haven't noticed it before but there are so many failures with any given commit I probably don't pay as much attention as I should to something that doesn't look related to the changes.  I will take a look now.

**Comment by lwhite1** [comment]
> > > CI shows a flake in testTable:
> > > ```
> > > Error:  org.apache.arrow.c.RoundtripTest.testTable  Time elapsed: 0.031 s  <<< ERROR!
> > > java.lang.IllegalStateException: Cannot import released ArrowSchema
> > > 	at org.apache.arrow.util.Preconditions.checkState(Preconditions.java:458)
> > > 	at org.apache.arrow.c.SchemaImporter.importField(SchemaImporter.java:63)
> > > 	at org.apache.arrow.c.SchemaImporter.importField(SchemaImporter.java:56)
> > > 	at org.apache.arrow.c.Data.importField(Data.java:246)
> > > 	at org.apache.arrow.c.Data.importSchema(Data.java:266)
> > > 	at org.apache.arrow.c.Data.importVectorSchemaRoot(Data.java:377)
> > > 	at org.apache.arrow.c.RoundtripTest.testTable(RoundtripTest.java:683)
> > > ```
> > 
> > 
> > @lwhite1 have you seen this? Might be worth some investigation
> 
> I haven't noticed it before but there are so many failures with any given commit I probably don't pay as much attention as I should to something that doesn't look related to the changes. I will take a look now.

It looks like a bug in the test. I will push a fix. IDK why it passes locally.

**Comment by lidavidm** [comment]
It looks like the test crashes on another attempt.

**Comment by lidavidm** [comment]
@davisusanibar can you rebase here to see if Larry's fix makes CI pass?

**Comment by lidavidm** [comment]
Ok, looks all good. Thanks!

**Pull request merged**